lib/ui/painting.dart

```diff
@@ -270,6 +270,43 @@ class Color {
     );
   }
 
+  /// Combine the foreground color as a transparent color over top
+  /// of a background color, and return the resulting combined color.
+  ///
+  /// This uses standard alpha blending ("SRC over DST") rules to produce a
+  /// blended color from two colors. This can be used as a performance
+  /// enhancement when trying to avoid needless alpha blending compositing
+  /// operations for two things that are solid colors with the same shape, but
+  /// overlay each other: instead, just paint one with the combined color.
+  static Color alphaBlend(Color foreground, Color background) {
+    final int alpha = foreground.alpha;
+    if (alpha == 0x00) { // Foreground completely transparent.
+      return background;
+    }
+    final int invAlpha = 0xff - alpha;
+    int backAlpha = background.alpha;
+    if (backAlpha == 0xff) { // Opaque background case
+      return new Color.fromARGB(
+        0xff,
+        (alpha * foreground.red + invAlpha * background.red) ~/ 0xff,
+        (alpha * foreground.green + invAlpha * background.green) ~/ 0xff,
+        (alpha * foreground.blue + invAlpha * background.blue) ~/ 0xff,
+      );
+    } else { // General case
+      backAlpha = (backAlpha * invAlpha) ~/ 0xff;
+      final int outAlpha = alpha + backAlpha;
+      if (outAlpha == 0x0) {
```

Comment:
alpha and backAlpha are always positive, since they are unsigned ints.
alpha is greater than zero, since you check for the zero case on line 283.
that means outAlpha must be greater than or equal to alpha.
which means outAlpha must be greater than zero.
so this line is dead code.

Reply:
oh hold on, I misread the code and missed that backAlpha wasn't just the alpha from a Color.

Reply:
ok yes:

alpha is in the range 1..255 (after line 285)
invAlpha is 255 - alpha, so it's in the range 0..254
backAlpha on line 295 is in the range 0..255
backAlpha on line 297 is therefore zero or positive, because it's the multiplication of two zero or positive values followed by an integer division (which can't change the sign)
therefore outAlpha is the addition of a number that is equal to or greater than 1, and a number that is equal to or greater than 0
therefore outAlpha cannot be 0.

Reply:
(i'd assert it)

Reply:
Thanks!  I added that check before I added the shortcut for the foreground alpha being zero, and forgot to go back and validate it.